kyo-prelude/shared/src/main/scala/kyo/Isolate.scala

```diff
@@ -0,0 +1,141 @@
+package kyo
+
+import internal.*
+import kyo.Flat
+import kyo.Frame
+import kyo.Kyo
+import kyo.Var.isolate
+import kyo.kernel.ContextEffect
+import kyo.kernel.Safepoint.Interceptor
+import scala.quoted.*
+
+/** Provides fine-grained control over state isolation and propagation within computations.
+  *
+  * The `Isolate` mechanism allows you to control how effect state is handled, contained, and propagated during computation. This is
+  * essential for implementing effects that need careful control over their state visibility and ordering of operations.
+  *
+  * While Isolate provides a low-level API through `use`, `resume`, and `restore` for advanced use cases, most users should rely on the
+  * high-level `run` method which composes these operations safely and handles all the complexity of state management internally.
+  *
+  * Effects typically provide their isolation strategies through their companion objects (e.g. `Var.isolate.update`, `Check.isolate.merge`).
+  * These isolates can be composed using `andThen`, which enforces the ordering of effect handling - ensuring effects are handled in the
+  * specified sequence.
+  *
+  * Key use cases include:
+  *   - Transactional effects where state changes may need to be contained until commit
+  *   - Effects that require explicit control over state propagation order
+  *   - Parallel computations where state needs to be merged in specific ways
+  *   - Scenarios where effect state should be temporarily scoped or discarded
+  *   - Error handling where partial effects must be rolled back atomically
+  *   - Complex workflows requiring nested transaction boundaries
+  *
+  * Implementations typically choose one of three strategies:
+  *   - Merge: Combine states when multiple computations need to reconcile their effects
+  *   - Update: Unconditionally updates the state with the last available value
+  *   - Discard: Prevent state from propagating beyond a defined scope
+  *
+  * A key feature of Isolate is its interaction with error handling through the Abort effect. When an Abort occurs within an isolated scope,
+  * the isolation acts as a transaction boundary - isolated effects within the isolated scope are rolled back or discarded, maintaining
+  * consistency of the outer scope. This transactional behavior applies across all effect types:
+  *   - State changes are rolled back to their pre-isolation values
+  *   - Emitted values are discarded rather than being merged
+  *   - Cached computations remain isolated and don't propagate
+  *   - Nested isolations maintain these guarantees at each level
+  *
+  * This combination of isolation strategies and error handling makes Isolate particularly powerful for implementing robust,
+  * transaction-like behaviors across different effect types while maintaining clean separation of concerns.
+  *
+  * @tparam S
+  *   The type of effect being isolated
+  */
+abstract class Isolate[S]:
```

Comment:
As discussed offline, I think it would be better if we could redesign this class to be an 'EffectScope' or 'Scope' etc. We should be able to then design a general `Scope.Strategy` to determine the strategy for managing Context effects within forks. I believe this would then enable broader application, combination, and more fluent operators.

I can imagine for most usecases, effects could define a typeclass to encode the required state handling for each 'strategy'. When necessary, we could ask the user to define something (Merge).

We could specialize the behavior on `Nothing`/`Any` to avoid the cost when no management within forks is required. Ideally this would also be merged with Boundary.


To avoid blocking this functionality, I will approve for now.

Reply:
Makes sense. I think having a more explicit mechanism without implicit resolution initially will be beneficial for us to identify patterns of how to safely make the mechanism more convenient. The trouble with inference is that effect handling ordering is important but we'll probably find usage patterns we can optimize for.